kaldb/src/main/java/com/slack/kaldb/server/ManagerApiGrpc.java

```diff
@@ -0,0 +1,206 @@
+
+      ServiceMetadata updatedServiceMetadata =
+          new ServiceMetadata(
+              serviceMetadata.getName(),
+              serviceMetadata.getOwner(),
+              updatedThroughputBytes,
+              updatedServicePartitionMetadata);
+      serviceMetadataStore.updateSync(updatedServiceMetadata);
+
+      responseObserver.onNext(
+          ManagerApi.UpdatePartitionAssignmentResponse.newBuilder()
+              .addAllAssignedPartitionIds(request.getPartitionIdsList())
+              .build());
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      LOG.error("Error updating partition assignment", e);
+      responseObserver.onError(Status.UNKNOWN.withDescription(e.getMessage()).asException());
+    }
+  }
+
+  /**
+   * Returns a new list of service partition metadata, with the provided partition IDs as the
+   * current active assignment. This finds the current active assignment (end time of max long),
+   * sets it to the current time, and then appends a new service partition assignment starting from
+   * current time + 1 to max long.
+   */
+  private static ImmutableList<ServicePartitionMetadata> addNewPartition(
+      List<ServicePartitionMetadata> existingPartitions, List<String> newPartitionIdsList) {
+    if (newPartitionIdsList.isEmpty()) {
+      return ImmutableList.copyOf(existingPartitions);
+    }
+
+    Optional<ServicePartitionMetadata> previousActiveServicePartition =
+        existingPartitions
+            .stream()
+            .filter(
+                servicePartitionMetadata ->
+                    servicePartitionMetadata.getEndTimeEpochMs() == Long.MAX_VALUE)
+            .findFirst();
+
+    List<ServicePartitionMetadata> remainingServicePartitions =
+        existingPartitions
+            .stream()
+            .filter(
+                servicePartitionMetadata ->
+                    servicePartitionMetadata.getEndTimeEpochMs() != Long.MAX_VALUE)
+            .collect(Collectors.toList());
+
+    // todo - consider adding some padding to this value; this may complicate
+    //   validation as you would need to consider what happens when there's a future
+    //   cut-over already scheduled
+    long partitionCutoverTime = Instant.now().toEpochMilli();
+
+    ImmutableList.Builder<ServicePartitionMetadata> builder =
+        ImmutableList.<ServicePartitionMetadata>builder().addAll(remainingServicePartitions);
+
+    if (previousActiveServicePartition.isPresent()) {
+      ServicePartitionMetadata updatedPreviousActivePartition =
+          new ServicePartitionMetadata(
+              previousActiveServicePartition.get().getStartTimeEpochMs(),
+              partitionCutoverTime,
+              previousActiveServicePartition.get().getPartitions());
+      builder.add(updatedPreviousActivePartition);
+    }
+
+    ServicePartitionMetadata newPartitionMetadata =
+        new ServicePartitionMetadata(partitionCutoverTime + 1, Long.MAX_VALUE, newPartitionIdsList);
```

Comment:
I think it's best to pass in the padding as a parameter to this method. That way the implementation would be oblivious of the padding value. The caller can decide what value to set it as.

Also, please use the static variable instead of directly calling Long.MAX_VALUE.

Reply:
Added static variable for the long max value. Updated comment for when we do support padding that it should be exposed as param.